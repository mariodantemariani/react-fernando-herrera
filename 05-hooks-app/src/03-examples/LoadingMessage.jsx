export const LoadingMessage = () => {
  return (
    <section
      style={{ textAlign: "center", height: "200px" }}
      className="alert alert-info text-center p-0 m-0 d-flex align-items-center justify-content-center"
    >
      Cargando...
    </section>
  );
};
