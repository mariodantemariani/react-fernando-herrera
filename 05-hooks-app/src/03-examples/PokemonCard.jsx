import { PropTypes } from "prop-types";
import { useLayoutEffect, useRef, useState } from "react";

export const PokemonCard = ({ id, name, sprites = [] }) => {
  const pRef = useRef();

  const [boxSize, setBoxSize] = useState({
    width: 0,
    height: 0,
  });

  useLayoutEffect(() => {
    const { height, width } = pRef.current.getBoundingClientRect();
    setBoxSize({ height, width });
  }, [name]);

  return (
    <section style={{ height: "200px" }}>
      <h2 className="text-capitalize" ref={pRef}>
        #{id} - {name}
      </h2>

      <div>
        {sprites.map((sprite) => (
          <img key={sprite} src={sprite} alt={name} />
        ))}
      </div>

      <code>{JSON.stringify({ boxSize })}</code>
    </section>
  );
};

PokemonCard.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  sprites: PropTypes.arrayOf(PropTypes.string),
};
