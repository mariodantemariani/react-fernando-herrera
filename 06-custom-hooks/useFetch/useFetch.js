import { useState, useEffect } from "react";

const localCache = {};

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    isLoading: true,
    hasError: null,
    error: null,
  });

  useEffect(() => {
    getFetch();
  }, [url]);

  const setLoadingState = () => {
    setState({
      data: null,
      isLoading: true,
      hasError: false,
      error: null,
    });
  };

  const getFetch = async () => {
    if (localCache[url]) {
      console.log("from cache");
      setState({
        data: localCache[url],
        isLoading: false,
        hasError: false,
        error: null,
      });
      return;
    }

    setLoadingState();

    const resp = await fetch(url);
    await new Promise((resolve) => setTimeout(resolve, 1500));

    if (!resp.ok) {
      setState({
        data: null,
        isLoading: false,
        hasError: true,
        error: {
          status: resp.status,
          message: "Error trayendo la data",
        },
      });
      return;
    }

    const data = await resp.json();

    setState({
      data,
      isLoading: false,
      hasError: false,
      error: null,
    });

    localCache[url] = data;
  };

  return {
    data: state.data,
    isLoading: state.isLoading,
    hasError: state.hasError,
    error: state.error,
  };
};
