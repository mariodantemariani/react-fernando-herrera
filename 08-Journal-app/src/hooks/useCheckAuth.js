import { useDispatch, useSelector } from "react-redux";

import { useEffect } from "react";
import { onAuthStateChanged } from "firebase/auth";
import { FirebaseAuth } from "../firebase/config";
import { login, logout } from "../store/auth";
import { startLoadingNotes } from "../store/journal";

export const useCheckAuth = () => {
  console.log("useCheckAuth");
  const { status } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  //disparar un efecto para disparar si la persona esta autenticada o no.
  //si esta autenticada redirigir a al main o si no redirigir a Auth
  //usando firebase
  useEffect(() => {
    onAuthStateChanged(FirebaseAuth, async (user) => {
      if (!user) return dispatch(logout());

      const { uid, displayName, email, photoURL } = user;

      dispatch(login({ uid, displayName, email, photoURL }));
      dispatch(startLoadingNotes());
    });
  }, [status]);

  return status;
};
