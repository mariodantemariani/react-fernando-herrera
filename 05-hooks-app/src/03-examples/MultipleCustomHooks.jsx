import { useFetch, useCounter } from "../hooks";
import { LoadingMessage } from "./LoadingMessage";
import { PokemonCard } from "./PokemonCard";

export const MultipleCustomHooks = () => {
  const { counter, increment, decrement } = useCounter(1);

  const url = `https://pokeapi.co/api/v2/pokemon/${counter}/`;
  const { data, isLoading, hasError, error } = useFetch(url);

  return (
    <>
      MultipleCustomHooks
      <h1>Informacion del pokemon</h1>
      <hr />
      {/* {isLoading ? <LoadingMessage /> : <PokemonCard {...data} />} */}
      {isLoading ? (
        <LoadingMessage />
      ) : (
        <PokemonCard
          id={data?.id}
          name={data?.name}
          sprites={[
            data.sprites?.front_default,
            data.sprites?.front_shiny,
            data.sprites?.back_default,
            data.sprites?.back_shiny,
          ]}
        />
      )}
      {hasError && <div>{error.message}</div>}
      <button
        disabled={counter === 1}
        className="btn btn-primary mt-2"
        onClick={() => (counter > 1 ? decrement(1) : null)}
      >
        Anterior
      </button>
      <button className="btn btn-primary mt-2" onClick={() => increment(1)}>
        Siguiente
      </button>
    </>
  );
};
