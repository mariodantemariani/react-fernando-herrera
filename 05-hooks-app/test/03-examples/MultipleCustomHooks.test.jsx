import { render, screen, fireEvent } from "@testing-library/react";

import { MultipleCustomHooks } from "../../src/03-examples/MultipleCustomHooks";

import { useFetch } from "../../src/hooks/useFetch";

import { useCounter } from "../../src/hooks/useCounter";

jest.mock("../../src/hooks/useCounter");
jest.mock("../../src/hooks/useFetch");

describe("Prueba en MultipleCustomHooks", () => {
  const mockIncrement = jest.fn();
  const mockDecrement = jest.fn();

  useCounter.mockReturnValue({
    counter: 1,
    increment: mockIncrement,
    decrement: mockDecrement,
  });

  beforeEach(() => jest.clearAllMocks());

  test("should be the same that snapShot", () => {
    useFetch.mockReturnValue({
      data: null,
      isLoading: true,
      hasError: null,
    });

    const { container } = render(<MultipleCustomHooks />);
    expect(container).toMatchSnapshot();
  });

  test("should render default component", () => {
    useFetch.mockReturnValue({
      data: null,
      isLoading: true,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    expect(screen.getByText("Cargando...")).toBeTruthy();
    expect(screen.getByText("MultipleCustomHooks")).toBeTruthy();

    const nextButton = screen.getByRole("button", { name: "Siguiente" });
    expect(nextButton.disabled).not.toBeTruthy();

    const previousButton = screen.getByRole("button", { name: "Anterior" });
    expect(previousButton.disabled).toBeTruthy();
  });

  test("should render pokemon card", () => {
    useFetch.mockReturnValue({
      data: {
        id: 1,
        name: "Bulbasaur",
        sprites: {
          front_default:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          front_shiny:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/shiny/1.png",
          back_default:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
          back_shiny:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/shiny/1.png",
        },
      },
      error: null,
      isLoading: false,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    const text = "Bulbasaur";
    expect(screen.getByText(text)).toBeTruthy();

    const previousButton = screen.getByRole("button", { name: "Anterior" });
    expect(previousButton.disabled).toBeTruthy();
    const nextButton = screen.getByRole("button", { name: "Siguiente" });
    expect(nextButton.disabled).not.toBeTruthy();
    //    screen.debug();
  });

  test("Should call increment and decrement", () => {
    useFetch.mockReturnValue({
      data: {
        id: 1,
        name: "Bulbasaur",
        sprites: {
          front_default:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          front_shiny:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/shiny/1.png",
          back_default:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
          back_shiny:
            "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/shiny/1.png",
        },
      },
      error: null,
      isLoading: false,
      hasError: null,
    });

    render(<MultipleCustomHooks />);

    const nextButton = screen.getByRole("button", { name: "Siguiente" });
    fireEvent.click(nextButton);

    expect(mockIncrement).toHaveBeenCalled();
  });
});
