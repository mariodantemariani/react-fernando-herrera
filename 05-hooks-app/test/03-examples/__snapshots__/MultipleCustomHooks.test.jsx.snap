// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en MultipleCustomHooks should be the same that snapShot 1`] = `
<div>
  MultipleCustomHooks
  <h1>
    Informacion del pokemon
  </h1>
  <hr />
  <section
    class="alert alert-info text-center p-0 m-0 d-flex align-items-center justify-content-center"
    style="text-align: center; height: 200px;"
  >
    Cargando...
  </section>
  <button
    class="btn btn-primary mt-2"
    disabled=""
  >
    Anterior
  </button>
  <button
    class="btn btn-primary mt-2"
  >
    Siguiente
  </button>
</div>
`;
